File: cane-cpp/cane-core/src/geometry/clipping.cpp
#include "clipping.hpp"
#include "cane/error.hpp"
#include <algorithm>
#include <cmath>
#include <limits>
#include <numeric>

namespace cane::geometry {
namespace {
[[noreturn]] void limit() { throw Error(ErrorCode::resource_limit, "apply", "Clipped attachment exceeds the v1 geometry budget."); }
float cross(const Vertex& a, const Vertex& b, const Vertex& p) { return (b.x - a.x) * (p.y - a.y) - (b.y - a.y) * (p.x - a.x); }
float area(const Polygon& p) {
    float sum = 0;
    for (std::size_t i = 0; i < p.size(); ++i) { const auto& a = p[i]; const auto& b = p[(i + 1) % p.size()]; sum += a.x * b.y - b.x * a.y; }
    return sum * .5f;
}
bool valid(const Polygon& p) { const auto a = area(p); return p.size() >= 3 && std::isfinite(a) && std::abs(a) > clip_epsilon; }
bool equal(const Vertex& a, const Vertex& b, float epsilon, bool uv) {
    return std::abs(a.x - b.x) <= epsilon && std::abs(a.y - b.y) <= epsilon
        && (!uv || (std::abs(a.u - b.u) <= epsilon && std::abs(a.v - b.v) <= epsilon));
}
void push(Polygon& p, const Vertex& v) { if (p.size() >= max_render_attachment_vertices_v1) limit(); p.push_back(v); }
Vertex lerp(const Vertex& a, const Vertex& b, float t) {
    const Vertex result{a.x + (b.x - a.x) * t, a.y + (b.y - a.y) * t, a.u + (b.u - a.u) * t, a.v + (b.v - a.v) * t};
    if (!std::isfinite(result.x) || !std::isfinite(result.y) || !std::isfinite(result.u) || !std::isfinite(result.v))
        throw Error(ErrorCode::non_finite, "apply", "Clipping intersection is non-finite.");
    return result;
}
template<class Distance> void plane(const Polygon& input, Polygon& output, Distance distance, float epsilon, bool uv) {
    output.clear(); if (input.empty()) return;
    auto previous = input.back(); float previous_distance = distance(previous); bool previous_inside = previous_distance >= -epsilon;
    const auto append = [&](const Vertex& v) { if (output.empty() || !equal(output.back(), v, epsilon, uv)) push(output, v); };
    for (const auto& current : input) {
        const float current_distance = distance(current); const bool inside = current_distance >= -epsilon;
        if (!std::isfinite(previous_distance) || !std::isfinite(current_distance)) throw Error(ErrorCode::non_finite, "apply", "Clipping half-plane is non-finite.");
        if (inside != previous_inside) {
            const float denominator = previous_distance - current_distance;
            if (!std::isfinite(denominator)) throw Error(ErrorCode::non_finite, "apply", "Clipping denominator is non-finite.");
            if (std::abs(denominator) > matrix_epsilon) append(lerp(previous, current, std::clamp(previous_distance / denominator, 0.0f, 1.0f)));
        }
        if (inside) append(current);
        previous = current; previous_distance = current_distance; previous_inside = inside;
    }
    if (output.size() > 1 && equal(output.front(), output.back(), epsilon, uv)) output.pop_back();
}
Polygon& pool_item(std::vector<Polygon>& pool, std::size_t index) {
    if (index >= max_render_attachment_vertices_v1) limit();
    if (index == pool.size()) pool.emplace_back();
    return pool[index];
}
void append_piece(PreparedClip& output, const Polygon& piece) { pool_item(output.pieces, output.piece_count++) = piece; }
void triangle(const GeometryBuffer& source, std::size_t offset, Polygon& output) {
    output.clear();
    for (std::size_t i = 0; i < 3; ++i) output.push_back(source.vertices.at(source.indices.at(offset + i)));
}
void intersect(Polygon& a, Polygon& b, const Polygon& clip) {
    for (std::size_t i = 0; i < clip.size(); ++i) {
        const auto& start = clip[i]; const auto& end = clip[(i + 1) % clip.size()];
        plane(a, b, [&](const Vertex& p) { return cross(start, end, p); }, clip_epsilon, false); a.swap(b);
        if (!valid(a)) { a.clear(); return; }
    }
}
}
void GeometryBuffer::append(const Polygon& polygon) {
    if (polygon.size() < 3) return;
    if (polygon.size() > max_render_attachment_vertices_v1 || vertices.size() > max_render_attachment_vertices_v1 - polygon.size()) limit();
    const auto triangles = polygon.size() - 2;
    if (triangles > (indices.max_size() - indices.size()) / 3) limit();
    const auto offset = static_cast<std::uint32_t>(vertices.size()); vertices.insert(vertices.end(), polygon.begin(), polygon.end());
    for (std::uint32_t i = 1; i + 1 < polygon.size(); ++i) { indices.push_back(offset); indices.push_back(offset + i); indices.push_back(offset + i + 1); }
}
void prepare_clip(const std::vector<float>& xy, bool convex, bool inverse, std::optional<std::size_t> end_slot, PreparedClip& output, ClipWorkspace& scratch) {
    output.clear(); auto& points = scratch.points; points.clear();
    if (xy.size() % 2 != 0) throw Error(ErrorCode::validation_failed, "apply", "Clipping coordinate count must be even.");
    if (xy.size() / 2 > max_render_attachment_vertices_v1) limit();
    for (std::size_t i = 0; i < xy.size(); i += 2) {
        if (!std::isfinite(xy[i]) || !std::isfinite(xy[i + 1])) return; // A sampled invalid clip replaces the previous active clip.
        points.push_back({xy[i], xy[i + 1], 0, 0});
    }
    const auto signed_area = area(points); if (!valid(points)) return;
    if (signed_area < 0) std::reverse(points.begin(), points.end());
    bool is_convex = true;
    for (std::size_t i = 0; i < points.size(); ++i) if (cross(points[i], points[(i + 1) % points.size()], points[(i + 2) % points.size()]) < -clip_epsilon) { is_convex = false; break; }
    if (is_convex) append_piece(output, points);
    else if (convex) {
        std::sort(points.begin(), points.end(), [](const Vertex& a, const Vertex& b) { return a.x == b.x ? a.y < b.y : a.x < b.x; });
        points.erase(std::unique(points.begin(), points.end(), [](const Vertex& a, const Vertex& b) { return a.x == b.x && a.y == b.y; }), points.end());
        auto& hull = scratch.hull; hull.clear();
        for (const auto& p : points) { while (hull.size() >= 2 && cross(hull[hull.size() - 2], hull.back(), p) <= clip_epsilon) hull.pop_back(); hull.push_back(p); }
        const auto lower = hull.size();
        for (std::size_t i = points.size() - 1; i-- > 0;) {
            const auto& p = points[i]; while (hull.size() > lower && cross(hull[hull.size() - 2], hull.back(), p) <= clip_epsilon) hull.pop_back(); hull.push_back(p);
        }
        if (!hull.empty()) hull.pop_back();
        if (!valid(hull)) return;
        append_piece(output, hull);
    } else {
        auto& remaining = scratch.ear_indices; remaining.resize(points.size()); std::iota(remaining.begin(), remaining.end(), std::size_t{0});
        while (remaining.size() > 3) {
            bool found = false;
            for (std::size_t i = 0; i < remaining.size(); ++i) {
                const auto prev = remaining[(i + remaining.size() - 1) % remaining.size()], here = remaining[i], next = remaining[(i + 1) % remaining.size()];
                const auto& a = points[prev]; const auto& b = points[here]; const auto& c = points[next];
                if (cross(a, b, c) <= clip_epsilon) continue;
                bool contains = false;
                for (const auto index : remaining) {
                    if (index == prev || index == here || index == next) continue;
                    const auto& p = points[index];
                    if (cross(a, b, p) >= -clip_epsilon && cross(b, c, p) >= -clip_epsilon && cross(c, a, p) >= -clip_epsilon) { contains = true; break; }
                }
                if (contains) continue;
                auto& piece = pool_item(output.pieces, output.piece_count++); piece.assign({a, b, c});
                remaining.erase(remaining.begin() + static_cast<std::ptrdiff_t>(i)); found = true; break;
            }
            if (!found) { output.clear(); return; }
        }
        auto& piece = pool_item(output.pieces, output.piece_count++); piece.clear(); for (const auto i : remaining) piece.push_back(points[i]);
    }
    output.active = true; output.inverse = inverse; output.end_slot = end_slot;
}
void apply_clip(const GeometryBuffer& source, const PreparedClip& clip, GeometryBuffer& output, ClipWorkspace& scratch) {
    output.clear(); auto& a = scratch.a; auto& b = scratch.b;
    for (std::size_t t = 0; t < source.indices.size(); t += 3) {
        if (!clip.inverse) {
            for (std::size_t i = 0; i < clip.piece_count; ++i) { triangle(source, t, a); intersect(a, b, clip.pieces[i]); if (valid(a)) output.append(a); }
        } else {
            triangle(source, t, pool_item(scratch.fragments, 0)); std::size_t count = 1;
            for (std::size_t piece = 0; piece < clip.piece_count && count != 0; ++piece) {
                std::size_t next_count = 0, vertices = 0; const auto& polygon = clip.pieces[piece];
                for (std::size_t f = 0; f < count; ++f) {
                    a = scratch.fragments[f];
                    for (std::size_t edge = 0; edge < polygon.size(); ++edge) {
                        const auto& start = polygon[edge]; const auto& end = polygon[(edge + 1) % polygon.size()];
                        plane(a, b, [&](const Vertex& p) { return -cross(start, end, p); }, clip_epsilon, false);
                        if (valid(b)) {
                            if (vertices > max_render_attachment_vertices_v1 - b.size()) limit();
                            vertices += b.size(); pool_item(scratch.next_fragments, next_count++) = b;
                        }
                        plane(a, b, [&](const Vertex& p) { return cross(start, end, p); }, clip_epsilon, false); a.swap(b);
                        if (!valid(a)) break;
                    }
                }
                scratch.fragments.swap(scratch.next_fragments); count = next_count;
            }
            for (std::size_t f = 0; f < count; ++f) if (valid(scratch.fragments[f])) output.append(scratch.fragments[f]);
        }
    }
}
bool trim_atlas(const GeometryBuffer& source, float min_u, float max_u, float min_v, float max_v, GeometryBuffer& output, ClipWorkspace& scratch) {
    bool inside = true;
    for (const auto& v : source.vertices) if (v.u < min_u - atlas_source_trim_epsilon || v.u > max_u + atlas_source_trim_epsilon
        || v.v < min_v - atlas_source_trim_epsilon || v.v > max_v + atlas_source_trim_epsilon) { inside = false; break; }
    if (inside) return false;
    output.clear(); auto& a = scratch.a; auto& b = scratch.b;
    for (std::size_t i = 0; i < source.indices.size(); i += 3) {
        triangle(source, i, a);
        plane(a, b, [&](const Vertex& v) { return v.u - min_u; }, atlas_source_trim_epsilon, true); a.swap(b);
        plane(a, b, [&](const Vertex& v) { return max_u - v.u; }, atlas_source_trim_epsilon, true); a.swap(b);
        plane(a, b, [&](const Vertex& v) { return v.v - min_v; }, atlas_source_trim_epsilon, true); a.swap(b);
        plane(a, b, [&](const Vertex& v) { return max_v - v.v; }, atlas_source_trim_epsilon, true); a.swap(b);
        output.append(a); // Atlas trim retains finite zero-area triangles, unlike world clip fragments.
    }
    return true;
}
}

File: cane-csharp/cane-core/Runtime/Geometry/Clipping.cs
using System;
using System.Collections.Generic;

namespace Cane.Geometry
{
    internal readonly struct Vertex
    {
        internal readonly float X, Y, U, V;
        internal Vertex(float x, float y, float u = 0, float v = 0) { X = x; Y = y; U = u; V = v; }
        internal static Vertex Lerp(Vertex a, Vertex b, float t) => new Vertex(a.X + (b.X - a.X) * t, a.Y + (b.Y - a.Y) * t, a.U + (b.U - a.U) * t, a.V + (b.V - a.V) * t);
    }
    internal sealed class GeometryBuffer
    {
        internal readonly List<Vertex> Vertices = new List<Vertex>();
        internal readonly List<int> Indices = new List<int>();
        internal void Append(List<Vertex> polygon)
        {
            if (polygon.Count < 3) return;
            if (Vertices.Count > 65536 - polygon.Count) throw new RuntimeException(RuntimeErrorCode.ResourceLimit, "apply", "Clipped attachment exceeds 65536 vertices.");
            int offset = Vertices.Count; Vertices.AddRange(polygon);
            for (int i = 1; i < polygon.Count - 1; i++) { Indices.Add(offset); Indices.Add(offset + i); Indices.Add(offset + i + 1); }
        }
    }
    internal sealed class ClipPolygon
    {
        internal readonly List<List<Vertex>> Pieces;
        internal readonly string? EndSlot;
        internal readonly bool Inverse;
        internal ClipPolygon(List<List<Vertex>> pieces, bool inverse, string? endSlot) { Pieces = pieces; Inverse = inverse; EndSlot = endSlot; }
    }
    internal static class Clipping
    {
        private const float Epsilon = 0.00001f;
        private static float Cross(Vertex a, Vertex b, Vertex p) => (b.X - a.X) * (p.Y - a.Y) - (b.Y - a.Y) * (p.X - a.X);
        private static float Area(List<Vertex> points)
        {
            float sum = 0; for (int i = 0; i < points.Count; i++) { var a = points[i]; var b = points[(i + 1) % points.Count]; sum += a.X * b.Y - b.X * a.Y; } return sum * 0.5f;
        }
        private static bool Valid(List<Vertex> p) => p.Count >= 3 && Math.Abs(Area(p)) > Epsilon;
        internal static ClipPolygon? Prepare(float[] xy, bool convex, bool inverse, string? endSlot)
        {
            var points = new List<Vertex>(); for (int i = 0; i < xy.Length; i += 2) points.Add(new Vertex(xy[i], xy[i + 1]));
            float area = Area(points); if (points.Count < 3 || !Numeric.Finite(area) || Math.Abs(area) <= Epsilon) return null;
            if (area < 0) points.Reverse(); bool isConvex = true;
            for (int i = 0; i < points.Count; i++) if (Cross(points[i], points[(i + 1) % points.Count], points[(i + 2) % points.Count]) < -Epsilon) { isConvex = false; break; }
            if (convex && !isConvex) { points = Hull(points); isConvex = true; }
            if (isConvex) return Valid(points) ? new ClipPolygon(new List<List<Vertex>> { points }, inverse, endSlot) : null;
            var remaining = new List<Vertex>(points); var pieces = new List<List<Vertex>>(); int budget = points.Count * points.Count;
            while (remaining.Count > 3)
            {
                bool found = false;
                for (int i = 0; i < remaining.Count; i++)
                {
                    int previous = (i + remaining.Count - 1) % remaining.Count, next = (i + 1) % remaining.Count;
                    Vertex a = remaining[previous], b = remaining[i], c = remaining[next]; if (Cross(a, b, c) <= Epsilon) continue;
                    bool inside = false;
                    for (int j = 0; j < remaining.Count; j++)
                    { if (j == previous || j == i || j == next) continue; Vertex p = remaining[j]; if (Cross(a, b, p) >= -Epsilon && Cross(b, c, p) >= -Epsilon && Cross(c, a, p) >= -Epsilon) { inside = true; break; } }
                    if (inside) continue;
                    pieces.Add(new List<Vertex> { a, b, c }); remaining.RemoveAt(i); found = true; break;
                }
                if (!found || --budget == 0) return null;
            }
            if (Valid(remaining)) pieces.Add(remaining);
            return new ClipPolygon(pieces, inverse, endSlot);
        }
        private static List<Vertex> Hull(List<Vertex> input)
        {
            var sorted = new List<Vertex>(input); sorted.Sort((a, b) => a.X == b.X ? a.Y.CompareTo(b.Y) : a.X.CompareTo(b.X));
            for (int i = sorted.Count - 1; i > 0; i--) if (sorted[i].X == sorted[i - 1].X && sorted[i].Y == sorted[i - 1].Y) sorted.RemoveAt(i);
            var lower = new List<Vertex>(); var upper = new List<Vertex>();
            foreach (var p in sorted) { while (lower.Count >= 2 && Cross(lower[lower.Count - 2], lower[lower.Count - 1], p) <= Epsilon) lower.RemoveAt(lower.Count - 1); lower.Add(p); }
            for (int i = sorted.Count - 1; i >= 0; i--) { var p = sorted[i]; while (upper.Count >= 2 && Cross(upper[upper.Count - 2], upper[upper.Count - 1], p) <= Epsilon) upper.RemoveAt(upper.Count - 1); upper.Add(p); }
            if (lower.Count != 0) lower.RemoveAt(lower.Count - 1); if (upper.Count != 0) upper.RemoveAt(upper.Count - 1); lower.AddRange(upper); return lower;
        }
        private static List<Vertex> Plane(List<Vertex> input, Func<Vertex, float> distance, float epsilon, bool compareUv)
        {
            var output = new List<Vertex>(); if (input.Count == 0) return output;
            Vertex previous = input[input.Count - 1]; float pd = distance(previous); bool pi = pd >= -epsilon;
            foreach (Vertex current in input)
            {
                float cd = distance(current); bool ci = cd >= -epsilon;
                if (ci != pi)
                {
                    float denominator = pd - cd;
                    if (Math.Abs(denominator) > Numeric.Epsilon) output.Add(Vertex.Lerp(previous, current, Numeric.Clamp(pd / denominator, 0, 1)));
                }
                if (ci) output.Add(current);
                previous = current; pd = cd; pi = ci;
            }
            for (int i = output.Count - 1; i > 0; i--) if (Same(output[i], output[i - 1], epsilon, compareUv)) output.RemoveAt(i);
            if (output.Count > 1 && Same(output[0], output[output.Count - 1], epsilon, compareUv)) output.RemoveAt(output.Count - 1);
            return output;
        }
        private static bool Same(Vertex a, Vertex b, float epsilon, bool uv) => Math.Abs(a.X - b.X) <= epsilon && Math.Abs(a.Y - b.Y) <= epsilon && (!uv || Math.Abs(a.U - b.U) <= epsilon && Math.Abs(a.V - b.V) <= epsilon);
        private static List<Vertex> Intersect(List<Vertex> input, List<Vertex> clip)
        {
            foreach (var edge in Edges(clip)) { input = Plane(input, p => Cross(edge.A, edge.B, p), Epsilon, false); if (!Valid(input)) return new List<Vertex>(); }
            return input;
        }
        private static IEnumerable<(Vertex A, Vertex B)> Edges(List<Vertex> points)
        { for (int i = 0; i < points.Count; i++) yield return (points[i], points[(i + 1) % points.Count]); }
        private static List<List<Vertex>> Subtract(List<Vertex> input, List<Vertex> clip)
        {
            var fragments = new List<List<Vertex>>();
            foreach (var edge in Edges(clip))
            {
                var outside = Plane(input, p => -Cross(edge.A, edge.B, p), Epsilon, false); if (Valid(outside)) fragments.Add(outside);
                input = Plane(input, p => Cross(edge.A, edge.B, p), Epsilon, false); if (!Valid(input)) break;
            }
            return fragments;
        }
        internal static GeometryBuffer Apply(GeometryBuffer source, ClipPolygon clip)
        {
            var output = new GeometryBuffer();
            for (int triangle = 0; triangle < source.Indices.Count; triangle += 3)
            {
                var original = new List<Vertex> { source.Vertices[source.Indices[triangle]], source.Vertices[source.Indices[triangle + 1]], source.Vertices[source.Indices[triangle + 2]] };
                if (!clip.Inverse) { foreach (var piece in clip.Pieces) { var p = Intersect(original, piece); if (Valid(p)) output.Append(p); } }
                else
                {
                    var fragments = new List<List<Vertex>> { original };
                    foreach (var piece in clip.Pieces)
                    {
                        var next = new List<List<Vertex>>(); foreach (var fragment in fragments) next.AddRange(Subtract(fragment, piece)); fragments = next;
                        long count = 0; foreach (var p in fragments) count += p.Count; if (count > 65536) throw new RuntimeException(RuntimeErrorCode.ResourceLimit, "apply", "Inverse clipping output exceeds vertex budget.");
                    }
                    foreach (var p in fragments) if (Valid(p)) output.Append(p);
                }
            }
            return output;
        }
        internal static GeometryBuffer Trim(GeometryBuffer source, float minU, float maxU, float minV, float maxV)
        {
            const float epsilon = 0.000001f; bool allInside = true;
            foreach (Vertex p in source.Vertices) if (p.U < minU - epsilon || p.U > maxU + epsilon || p.V < minV - epsilon || p.V > maxV + epsilon) { allInside = false; break; }
            if (allInside) return source;
            var output = new GeometryBuffer();
            for (int i = 0; i < source.Indices.Count; i += 3)
            {
                var p = new List<Vertex> { source.Vertices[source.Indices[i]], source.Vertices[source.Indices[i + 1]], source.Vertices[source.Indices[i + 2]] };
                p = Plane(p, v => v.U - minU, epsilon, true); p = Plane(p, v => maxU - v.U, epsilon, true);
                p = Plane(p, v => v.V - minV, epsilon, true); p = Plane(p, v => maxV - v.V, epsilon, true);
                if (p.Count >= 3) output.Append(p);
            }
            return output;
        }
    }
}
